limacharlie/firehose.go

```diff
@@ -0,0 +1,308 @@
+	})
+	certPrivKeyPEM := new(bytes.Buffer)
+	pem.Encode(certPrivKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(certPrivateKey),
+	})
+	cert, err := tls.X509KeyPair(certPEM.Bytes(), certPrivKeyPEM.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	return &cert, nil
+}
+
+func startListener(listenOnIP net.IP, listenOnPort uint16, sslCertPath string, sslCertKeyPath string) (*net.Listener, error) {
+	createTempCert := len(sslCertPath) == 0
+	createTempKey := len(sslCertKeyPath) == 0
+	if createTempCert && !createTempKey {
+		return nil, fmt.Errorf("certificate key path missing")
+	}
+	if !createTempCert && createTempKey {
+		return nil, fmt.Errorf("certificate path missing")
+	}
+
+	var certificate *tls.Certificate = nil
+	if createTempCert && createTempKey {
+		tempCert, err := createSelfSignedCertificate()
+		if err != nil {
+			return nil, fmt.Errorf("could not create self signed certificate: %s", err)
+		}
+		certificate = tempCert
+	} else {
+		tempCert, err := tls.LoadX509KeyPair(sslCertPath, sslCertKeyPath)
+		if err != nil {
+			return nil, fmt.Errorf("error loading certificate with cert path '%s' and key path '%s': %s", sslCertPath, sslCertKeyPath, err)
+		}
+		certificate = &tempCert
+	}
+
+	tlsConfig := tls.Config{
+		Certificates: []tls.Certificate{*certificate},
+		CipherSuites: []uint16{tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256},
+	}
+	tlsListener, err := tls.Listen("tcp", fmt.Sprintf("%s:%d", listenOnIP, listenOnPort), &tlsConfig)
+	if err != nil {
+		return nil, fmt.Errorf("could not start TLS listener: %s", err)
+	}
+	return &tlsListener, nil
+}
+
+func registerOutput(org Organization, outputOpts FirehoseOutputOptions) error {
+	if outputOpts.UniqueName == "" {
+		log.Info().Msg("output registration not required")
+		return nil
+	}
+
+	outputName := "tmp_live_" + outputOpts.UniqueName
+	allOutputs, err := org.Outputs()
+	if err != nil {
+		return fmt.Errorf("could not register output with name '%s': %s", outputName, err)
+	}
+
+	_, found := allOutputs[outputName]
+	if found {
+		log.Debug().Str("name", outputName).Msg("output registration already done")
+		return nil
+	}
+
+	output := makeGenericOutput(outputOpts)
+	_, err = org.OutputAdd(output)
+	if err != nil {
+		return fmt.Errorf("could not add output: %s", err)
+	}
+	log.Debug().Msg("output registration done")
+	return nil
+}
+
+func StartFirehose(org Organization, fhOpts FirehoseOptions) (Firehose, error) {
+	return StartFirehoseAndRegisterOutput(org, fhOpts, nil)
+}
+
+func StartFirehoseAndRegisterOutput(org Organization, fhOpts FirehoseOptions, fhOutputOpts *FirehoseOutputOptions) (Firehose, error) {
+	if fhOutputOpts != nil {
+		err := registerOutput(org, *fhOutputOpts)
+		if err != nil {
+			return Firehose{}, err
+		}
+	}
+
+	listener, err := startListener(fhOpts.ListenOnIP, fhOpts.ListenOnPort, fhOpts.SSLCertPath, fhOpts.SSLCertKeyPath)
+	if err != nil {
+		return Firehose{}, err
+	}
+
+	fh := Firehose{org,
+		fhOpts,
```

Comment:
For non-trivial structs, it's more readable to include the field names in the initialization, it also makes sure that if we add fields in the future we don't swap fields by mistake. So like:
```
fh := Firehose{
    Organization: org,
    opts: fhOpts,
    ...
}
```